src/audio/copier/copier.c

```diff
@@ -1098,6 +1120,62 @@ static int copier_unbind(struct processing_module *mod, void *data)
 	return 0;
 }
 
+static void mic_privacy_event(void *arg, enum notify_id type, void *data)
+{
+	LOG_INF("mic_privacy_event");
+	struct mic_privacy_data *mic_priv_data = arg;
+	struct mic_privacy_settings *mic_privacy_settings = data;
+
+	LOG_INF("mic_privacy_event, arg = %p, data = %p", arg, data);
+
+	if (type == NOTIFIER_ID_MIC_PRIVACY_STATE_CHANGE) {
+
+		LOG_INF("NOTIFIER_ID_MIC_PRIVACY_STATE_CHANGE, max ramp time = %d, ", mic_privacy_settings->max_ramp_time);
+		LOG_INF("mic_privacy_event, state1 = %d, state2 = %d ", mic_privacy_settings->mic_privacy_state, mic_priv_data->mic_privacy_state);
+
+		if (mic_privacy_settings->mic_privacy_state == UNMUTED) {
+			if (mic_priv_data->mic_privacy_state == MUTED) {
+				mic_priv_data->mic_privacy_state = FADE_IN;
+				LOG_INF("mic_privacy_event switch to FADE_IN");
+			}
+		} else {
+			//In case when mute would be triggered before copier instantiation.
+			if (mic_priv_data->mic_privacy_state != MUTED) {
+				mic_priv_data->mic_privacy_state = FADE_OUT;
+				LOG_INF("mic_privacy_event switch to FADE_OUT");
+			}
+		}
+		mic_priv_data->max_ramp_time_in_ms = (mic_privacy_settings->max_ramp_time * 1000) / ADSP_RTC_FREQUENCY;
+		LOG_INF("max_ramp_time_in_ms= %d, audio_freq = %d, max_ramp_time = %d", mic_priv_data->max_ramp_time_in_ms, mic_priv_data->audio_freq, mic_privacy_settings->max_ramp_time);
+
+	}
+}
+
+static int mic_privacy_configure(struct comp_dev *dev, struct copier_data *cd)
+{
+    int ret;
+    struct mic_privacy_data* mic_priv_data;
+    mic_priv_data = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(struct mic_privacy_data));
+    if (!mic_priv_data) {
+        rfree(mic_priv_data);
+        return -ENOMEM;
+    }
+
+    mic_priv_data->audio_freq =  cd->config.base.audio_fmt.sampling_frequency;
+
+    uint32_t zeroing_wait_time = (get_dma_zeroing_wait_time() *1000) / ADSP_RTC_FREQUENCY;
+
+    ret = copier_gain_set_params(dev, &mic_priv_data->mic_priv_gain_params, zeroing_wait_time, SOF_DAI_INTEL_NONE);
+    if(ret != 0)
+        return ret;
+
+    cd->mic_priv = mic_priv_data;
+
+    ret = notifier_register(cd->mic_priv, NULL, NOTIFIER_ID_MIC_PRIVACY_STATE_CHANGE,
+                      mic_privacy_event, 0);
```

Comment:
what if notifier registration fails? You return an error, will the component be usable after that? If not - also free memory